File: frontend/src/lib/habits.js
/**
 * Habit-tracking data layer (Phase 2).
 *
 * Patients write and read their own daily habit records directly through the
 * Firestore client SDK (guarded by security rules). A doctor's view goes through
 * the backend (`getPatientWellnessSummary`) so no client gets cross-user read
 * access.
 *
 * The aggregation here mirrors backend/app/services/habits.py exactly (same
 * metrics, targets, adherence formula, streak rule) so the patient dashboard and
 * the doctor/report views always tell the same story.
 *
 * One record per patient per local day, keyed by a deterministic doc id
 * `${uid}_${date}` so re-logging a day overwrites instead of duplicating:
 *   { patientId, date: 'YYYY-MM-DD', sleepHours, exerciseMinutes, waterGlasses,
 *     meditationMinutes, screenTimeHours, ts, updatedAt }
 */
import {
  collection,
  doc,
  onSnapshot,
  query,
  serverTimestamp,
  setDoc,
  where,
} from 'firebase/firestore'
import { db } from './firebase.js'

const DAY = 24 * 60 * 60 * 1000
export const DEFAULT_WINDOW_DAYS = 7

/** Canonical habit metrics — keep in lock-step with backend HABITS. */
export const HABITS = [
  { key: 'sleepHours', label: 'Sleep', unit: 'hrs', target: 8, higherIsBetter: true, icon: 'moon', max: 12, step: 0.5 },
  { key: 'exerciseMinutes', label: 'Exercise', unit: 'min', target: 30, higherIsBetter: true, icon: 'activity', max: 120, step: 5 },
  { key: 'waterGlasses', label: 'Water', unit: 'glasses', target: 8, higherIsBetter: true, icon: 'droplet', max: 16, step: 1 },
  { key: 'meditationMinutes', label: 'Meditation', unit: 'min', target: 10, higherIsBetter: true, icon: 'flower', max: 60, step: 1 },
  { key: 'screenTimeHours', label: 'Screen Time', unit: 'hrs', target: 6, higherIsBetter: false, icon: 'smartphone', max: 16, step: 0.5 },
]
export const HABIT_KEYS = HABITS.map((h) => h.key)

/** Local YYYY-MM-DD for a Date (defaults to now). */
export function isoDate(d = new Date()) {
  const y = d.getFullYear()
  const m = String(d.getMonth() + 1).padStart(2, '0')
  const day = String(d.getDate()).padStart(2, '0')
  return `${y}-${m}-${day}`
}

function num(v) {
  if (v === null || v === undefined || v === '') return null
  const n = Number(v)
  if (Number.isNaN(n)) return null
  return Math.max(0, n)
}

/** Per-metric adherence in [0,1] for a single day's value (null when unlogged). */
export function metricAdherence(spec, value) {
  if (value === null || value === undefined) return null
  if (spec.target <= 0) return 1
  if (spec.higherIsBetter) return Math.min(1, value / spec.target)
  if (value <= spec.target) return 1
  const over = (value - spec.target) / spec.target
  return Math.max(0, 1 - over)
}

/**
 * Upsert today's habit record. Pass only the metrics that changed; the rest are
 * merged. No-op without a patientId.
 */
export async function recordHabit(patientId, metrics = {}, date = isoDate()) {
  if (!patientId) return
  const clean = {}
  for (const key of HABIT_KEYS) {
    if (key in metrics) {
      const v = num(metrics[key])
      if (v !== null) clean[key] = v
    }
  }
  const ref = doc(db, 'habit_entries', `${patientId}_${date}`)
  await setDoc(
    ref,
    { patientId, date, ...clean, ts: Date.now(), updatedAt: serverTimestamp() },
    { merge: true },
  )
}

/** Live subscription to a patient's habit records. Returns the unsubscribe fn. */
export function subscribeHabits(patientId, onChange, onError) {
  const q = query(collection(db, 'habit_entries'), where('patientId', '==', patientId))
  return onSnapshot(
    q,
    (snap) => onChange(snap.docs.map((d) => d.data())),
    (err) => {
      console.error('[habits] subscription error', err)
      onError?.(err)
    },
  )
}

function dayRecords(entries) {
  const byDate = {}
  for (const e of entries || []) {
    if (!e.date) continue
    const prev = byDate[e.date]
    if (!prev || (e.ts || 0) >= (prev.ts || 0)) byDate[e.date] = e
  }
  return byDate
}

function parseDate(d) {
  const [y, m, dd] = (d || '').split('-').map(Number)
  if (!y || !m || !dd) return null
  return new Date(y, m - 1, dd)
}

function streak(dates, today) {
  if (!dates.size) return 0
  const set = new Set([...dates].map((d) => parseDate(d)?.getTime()).filter(Boolean))
  let cursor = parseDate(today) || new Date()
  const dayMs = DAY
  let t = cursor.getTime()
  if (!set.has(t)) {
    t -= dayMs // allow anchoring on yesterday if today isn't logged yet
    if (!set.has(t)) return 0
  }
  let count = 0
  while (set.has(t)) {
    count += 1
    t -= dayMs
  }
  return count
}

/**
 * Aggregate raw habit records into adherence / streak / consistency + per-metric
 * breakdown. Mirrors backend habits.summarize().
 * Returns { loggedDays, windowDays, adherence, streak, consistency, today, metrics }.
 */
export function aggregateHabits(entries, { now = Date.now(), windowDays = DEFAULT_WINDOW_DAYS, today = isoDate() } = {}) {
  windowDays = Math.max(1, windowDays)
  const cutoff = now - windowDays * DAY
  const byDate = dayRecords(entries)
  const inWindow = Object.values(byDate).filter((r) => (r.ts || 0) >= cutoff)
  const loggedDays = inWindow.length

  const metricAdhValues = []
  const sortedDesc = Object.values(byDate).sort((a, b) => (b.ts || 0) - (a.ts || 0))

  const metrics = HABITS.map((spec) => {
    const present = inWindow.map((r) => num(r[spec.key])).filter((v) => v !== null)
    const adh = present.map((v) => metricAdherence(spec, v)).filter((a) => a !== null)
    const avg = present.length ? Number((present.reduce((a, b) => a + b, 0) / present.length).toFixed(2)) : 0
    const mAdh = adh.length ? Number((adh.reduce((a, b) => a + b, 0) / adh.length).toFixed(4)) : 0
    if (adh.length) metricAdhValues.push(mAdh)

    let latest = null
    for (const r of sortedDesc) {
      const v = num(r[spec.key])
      if (v !== null) { latest = v; break }
    }
    return {
      key: spec.key, label: spec.label, unit: spec.unit, target: spec.target,
      higherIsBetter: spec.higherIsBetter, avg, adherence: mAdh,
      loggedDays: present.length, latest, onTrack: mAdh >= 0.7,
    }
  })

  const adherence = metricAdhValues.length
    ? Number((metricAdhValues.reduce((a, b) => a + b, 0) / metricAdhValues.length).toFixed(4))
    : 0
  const consistency = Number(Math.min(1, loggedDays / windowDays).toFixed(4))
  const todayRec = byDate[today] || null
  const todayValues = Object.fromEntries(HABITS.map((s) => [s.key, todayRec ? num(todayRec[s.key]) : null]))

  return {
    loggedDays, windowDays, adherence,
    streak: streak(new Set(Object.keys(byDate)), today),
    consistency, today: todayValues, metrics,
  }
}

export const HABIT_ICON_META = {
  sleepHours: '#818cf8',
  exerciseMinutes: '#34d399',
  waterGlasses: '#38bdf8',
  meditationMinutes: '#a78bfa',
  screenTimeHours: '#fb923c',
}

File: backend/app/services/habits.py
"""
Habit-tracking aggregation (Phase 2).

Pure, dependency-free functions that turn a list of daily habit records into
adherence scores, streaks, and consistency metrics. Kept free of any
Firestore/IO so it is trivially unit-testable; the API layer reads the raw
records via the Admin SDK and hands them here, and the wellness engine reuses
the same summariser. Mirrors :mod:`app.services.mood` in spirit and is kept in
lock-step with the frontend ``lib/habits.js`` (same targets, same scoring).

A habit record is a plain dict, one per patient per day:
    {
      "patientId": "...",
      "date": "2026-06-11",          # local YYYY-MM-DD (the canonical key)
      "sleepHours": 7.5,
      "exerciseMinutes": 30,
      "waterGlasses": 8,
      "meditationMinutes": 10,
      "screenTimeHours": 5,
      "ts": 1718000000000,            # client epoch-ms (ordering/windowing)
    }
Any missing metric is treated as "not logged" (None), never as zero, so an
unlogged day doesn't drag an adherence score down to 0.
"""
from __future__ import annotations

from dataclasses import dataclass

_DAY_MS = 24 * 60 * 60 * 1000

# Canonical habit metrics, their daily target, and the direction that counts as
# "good". ``higher_is_better=False`` means we want to stay *at or below* target
# (screen time). Targets are evidence-informed defaults, not medical advice.
@dataclass(frozen=True)
class HabitSpec:
    key: str
    label: str
    unit: str
    target: float
    higher_is_better: bool


HABITS: tuple[HabitSpec, ...] = (
    HabitSpec("sleepHours", "Sleep", "hrs", 8.0, True),
    HabitSpec("exerciseMinutes", "Exercise", "min", 30.0, True),
    HabitSpec("waterGlasses", "Water", "glasses", 8.0, True),
    HabitSpec("meditationMinutes", "Meditation", "min", 10.0, True),
    HabitSpec("screenTimeHours", "Screen Time", "hrs", 6.0, False),
)
HABIT_KEYS: tuple[str, ...] = tuple(h.key for h in HABITS)
_SPEC_BY_KEY: dict[str, HabitSpec] = {h.key: h for h in HABITS}

# Default analysis window (days) for adherence + consistency.
DEFAULT_WINDOW_DAYS = 7


def _num(value) -> float | None:
    """Coerce a logged metric to a non-negative float, or None when absent."""
    if value is None or value == "":
        return None
    try:
        v = float(value)
    except (TypeError, ValueError):
        return None
    return max(0.0, v)


def metric_adherence(spec: HabitSpec, value: float | None) -> float | None:
    """
    Per-metric adherence in [0, 1] for a single day's logged value.

    For "higher is better" metrics the ratio value/target is capped at 1.0. For
    "lower is better" (screen time) full credit at/under target, decaying to 0
    at twice the target. Returns None when the metric wasn't logged that day.
    """
    if value is None:
        return None
    if spec.target <= 0:
        return 1.0
    if spec.higher_is_better:
        return round(min(1.0, value / spec.target), 4)
    # Lower-is-better: 1.0 at/below target, linear down to 0 at 2x target.
    if value <= spec.target:
        return 1.0
    over = (value - spec.target) / spec.target
    return round(max(0.0, 1.0 - over), 4)


def _day_records(entries: list[dict]) -> dict[str, dict]:
    """Collapse to the latest record per ``date`` key (last write wins)."""
    by_date: dict[str, dict] = {}
    for e in entries:
        date = e.get("date")
        if not date:
            continue
        prev = by_date.get(date)
        if prev is None or e.get("ts", 0) >= prev.get("ts", 0):
            by_date[date] = e
    return by_date


def _streak(dates: set[str], today: str | None) -> int:
    """
    Count consecutive days (ending today or yesterday) that have a record.

    ``today`` is the local YYYY-MM-DD anchor. A streak survives if the user has
    logged today *or* yesterday (so the streak doesn't visually "break" simply
    because they haven't logged yet today).
    """
    if not dates:
        return 0
    from datetime import date as _date, timedelta

    def parse(d: str):
        try:
            y, m, dd = (int(x) for x in d.split("-"))
            return _date(y, m, dd)
        except (ValueError, AttributeError):
            return None

    anchor = parse(today) if today else None
    if anchor is None:
        # Fall back to the most recent logged date.
        parsed = sorted(filter(None, (parse(d) for d in dates)))
        if not parsed:
            return 0
        anchor = parsed[-1]

    parsed_dates = set(filter(None, (parse(d) for d in dates)))
    # Allow the streak to anchor on today or yesterday.
    cursor = anchor
    if cursor not in parsed_dates:
        cursor = anchor - timedelta(days=1)
        if cursor not in parsed_dates:
            return 0
    streak = 0
    while cursor in parsed_dates:
        streak += 1
        cursor = cursor - timedelta(days=1)
    return streak


def summarize(
    entries: list[dict],
    *,
    now_ms: int,
    window_days: int = DEFAULT_WINDOW_DAYS,
    today: str | None = None,
) -> dict:
    """
    Build the full habit summary from raw daily records.

    Returns:
        {
          "logged_days": int,                 # days with at least one metric
          "window_days": int,
          "adherence": float,                 # 0..1 overall, across the window
          "streak": int,                      # consecutive logged days
          "consistency": float,               # 0..1 = logged_days / window_days
          "today": { metric: value | None },  # today's logged values
          "metrics": [                         # per-metric breakdown
            { key, label, unit, target, higher_is_better,
              avg, adherence, logged_days, latest, on_track }
          ],
        }
    """
    window_days = max(1, int(window_days))
    cutoff = now_ms - window_days * _DAY_MS
    by_date = _day_records(entries)

    in_window = [r for r in by_date.values() if r.get("ts", 0) >= cutoff]
    logged_days = len(in_window)

    metrics = []
    metric_adh_values: list[float] = []
    for spec in HABITS:
        vals = [_num(r.get(spec.key)) for r in in_window]
        present = [v for v in vals if v is not None]
        adh = [metric_adherence(spec, v) for v in present]
        adh = [a for a in adh if a is not None]
        avg = round(sum(present) / len(present), 2) if present else 0.0
        m_adherence = round(sum(adh) / len(adh), 4) if adh else 0.0
        if adh:
            metric_adh_values.append(m_adherence)

        # Latest logged value for this metric (most recent date overall).
        latest = None
        for r in sorted(by_date.values(), key=lambda x: x.get("ts", 0), reverse=True):
            v = _num(r.get(spec.key))
            if v is not None:
                latest = v
                break

        metrics.append(
            {
                "key": spec.key,
                "label": spec.label,
                "unit": spec.unit,
                "target": spec.target,
                "higher_is_better": spec.higher_is_better,
                "avg": avg,
                "adherence": m_adherence,
                "logged_days": len(present),
                "latest": latest,
                "on_track": m_adherence >= 0.7,
            }
        )

    overall = round(sum(metric_adh_values) / len(metric_adh_values), 4) if metric_adh_values else 0.0
    consistency = round(min(1.0, logged_days / window_days), 4)
    streak = _streak(set(by_date.keys()), today)

    # Today's logged snapshot.
    today_rec = by_date.get(today) if today else None
    today_values = {
        spec.key: (_num(today_rec.get(spec.key)) if today_rec else None) for spec in HABITS
    }

    return {
        "logged_days": logged_days,
        "window_days": window_days,
        "adherence": overall,
        "streak": streak,
        "consistency": consistency,
        "today": today_values,
        "metrics": metrics,
    }


def habit_report_summary(entries: list[dict], *, now_ms: int) -> dict | None:
    """
    Condense habit history into short prose blocks for the clinical PDF / session
    report. Returns ``None`` when nothing has been logged.
    """
    summary = summarize(entries, now_ms=now_ms, window_days=DEFAULT_WINDOW_DAYS)
    if summary["logged_days"] == 0:
        return None

    parts = []
    for m in summary["metrics"]:
        if m["logged_days"] == 0:
            continue
        unit = m["unit"]
        parts.append(f"{m['label']} avg {m['avg']}{unit} (target {m['target']}{unit})")
    if not parts:
        return None

    return {
        "habit_summary": (
            f"Over the last {summary['window_days']} days the patient logged habits on "
            f"{summary['logged_days']} day(s) with an overall adherence of "
            f"{round(summary['adherence'] * 100)}% (current streak {summary['streak']} day(s))."
        ),
        "habit_breakdown": "; ".join(parts) + ".",
        "habit_score": round(summary["adherence"] * 100),
    }
